### PR TITLE
Improved fitness function

### PR DESCRIPTION
**Pull request by rasmusmarak**

### CODE REVIEW

**Comment by rasmusmarak** [comment]
Here is the issue regarding an improved terminating event function for collision with body. #39

**Review by gomezzz — CHANGES_REQUESTED** [review]
Please try to improve the structure of the code, I don't understand the code at all. You have way too many functions in places where they make no sense.

Ideally, add a folder fitness functions with the following files:
* fitness_function_enums.py - Contains an enum with the names for the different functions
* fitness_function.py - Contains for each of the above enums exactly one function that receives position and time as input and evaluates the fitness according to its scheme and function that returns the respective function
* fitness_function.utils.py - Contains utility functions for the above functions, e.g. estimating the volume covered by trajectory 

The UDP can import the necessary fitness function with something like

```py
from fitness.fitness_function_enums import FitnessFunctions
from fitness.fitness_functions import get_fitness_function

chosen_function = FitnessFunctions.CoveredVolumeFarDistancePenalty

fitness_function = get_fitness_function(chosen_function)

# and later you just call
fitness = fitness_function(position,timesteps)
```

**Comment by rasmusmarak** [comment]
I have now finished all of the requested changes as well as some general restructuring of the code. Given this, it may appear as a great number of "new changes and code" on the reviewing end, likely as a consequence of the moving of files to new folders and separation of functions into new files for a more logical structure. But hopefully it is a bit more well structured now, especially compared to before. The replacement of the previously used notebook is included in the other PR regarding parallelization (as this feedback concerned both PRs).

File: toss/fitness/fitness_functions.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    """ Computes average distance to target altitude for satellite positions along the trajectory.
+    Args:
+        args (dotmap.DotMap): Dotmap with information on target squared altitude of spacecraft.
+        positions (np.ndarray): A set of positions along the trajectory.
+        timesteps (None): Time values for each position. 
+
+    Returns:
+        fitness (float): Average distance to target altitude.
+    """    
+    fitness = np.mean(np.abs(positions[0,:]**2 + positions[1,:]**2 + positions[2,:]**2 -args.problem.target_squared_altitude))
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:35]: a helper function 

```py
def _compute_squared_distance(arr,constant):
    return np.abs(np.sum(np.pow(arr,2)) - constant)
```

could be useful in a bunch of places in this file

File: toss/fitness/fitness_functions.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    Args:
+        args (dotmap.DotMap): Dotmap with information on radius of outer bounding sphere.
+        positions (np.ndarray): A set of positions along the trajectory.
+        timesteps (None): Time values for each position. 
+    Returns:
+        fitness (float): Average distance to radius of outer bounding-sphere.
+    """
+    r = positions[0:3, :]
+    r = (r[0,:]**2 + r[1,:]**2 + r[2, :]**2) - args.problem.radius_outer_boundings_sphere
+    r[r<0] = 0 #removes all positions inside the measurement-zone.
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:67]: as above, would make it smooth

File: toss/fitness/fitness_functions.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    """Computes the ration of unmeasured volume.
+    Args:
+        args (dotmap.DotMap): Dotmap with information on total measurable volume.
+        positions (np.ndarray): A set of positions along the trajectory.
+        timesteps (None): Time values for each position. 
+
+    Returns:
+        measured_volume_ratio (float): Ratio of unmeasured volume.
+    """
+    from fitness_function_utils import estimate_covered_volume
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:82]: please put all imports at the top of the file

( https://peps.python.org/pep-0008/#imports )

File: toss/fitness/fitness_functions.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    """ Returns combined fitness value of CoveredVolume and FarDistancePenalty
+    Args:
+        args (dotmap.DotMap): Dotmap with information on total measurable volume and radius of outer bounding sphere.
+        positions (np.ndarray): A set of positions along the trajectory.
+        timesteps (None): Time values for each position. 
+
+    Returns:
+        (float): Aggregate fitness value.
+    """
+    return (CoveredVolume(args,positions,timesteps) + FarDistancePenalty(args,positions,timesteps))
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:98]: If I am not mistaken , the covered volume is 0 to 1 and the far distance penalty 0 to a lot, so you need to consider some weighting here. I think you can just weight covered volume based on `args.problem.radius_outer_boundings_sphere` probably? :)

File: toss/optimization/udp_initial_condition.py [hunk 5]
```diff
@@ -90,35 +98,45 @@ def __init__(self, args, lower_bounds, upper_bounds):
...
 
         Returns:
-            fitness value (_float_): Difference between squared values of current and target altitude of satellite.
+            fitness (_float_): Evaluated fitness for user-specified fitness-function.
         """
 
-        # Integrate trajectory
-        _, squared_altitudes, collision_detected = trajectory_tools.compute_trajectory(x, self.args, equations_of_motion.compute_motion)
+        # Compute trajectory
+        from trajectory.compute_trajectory import compute_trajectory
```
Comment (gomezzz) [inline toss/optimization/udp_initial_condition.py:111]: imports at the top pls

File: toss/trajectory/compute_trajectory.py [hunk 6]
```diff
@@ -51,11 +38,13 @@ def compute_trajectory(x: np.ndarray, args, func: Callable) -> Union[np.ndarray,
...
     Returns:
-        trajectory_info (np.ndarray): Numpy array containing information on position and velocity at every time step (columnwise).
-        squared_altitudes (float): Sum of squared altitudes above origin for every position
         collision_penalty (bool): Penalty value given for the event of a collision with the celestial body.
+        list_of_trajectory_objects (list): List of OdeSystem integration objects (provided by DEsolver)
+        integration_intervals (np.ndarray): (1,N) Array containing the discretized and integrated time intervals. 
+
     """    
     # Separate initial state from chromosome and translate from osculating elements to cartesian frame.
+    import pykep as pk
```
[the remark this answers is not in the document]
Comment (gomezzz) [reply toss/trajectory/compute_trajectory.py:47]: (if you want to automate my complaints you could consider using flake :) Just take line 1 - 24 from this https://github.com/aidotse/PASEOS/blob/main/.github/workflows/run_tests.yml#L12 )

File: toss/trajectory/compute_trajectory.py [hunk 7]
```diff
@@ -186,8 +169,13 @@ def integrate_system(func: Callable, x: np.ndarray, args):
                 time_of_maneuver (float): Time for adding impulsive maneuver [seconds].
                 delta_v (np.ndarray): Array containing the cartesian componants of the impulsive maneuver.
     Returns:
-        trajectory (desolver.differential_system.OdeSystem): The integration object provided by desolver.
+        ode_object (desolver.differential_system.OdeSystem): The integration object provided by desolver.
     """
+    # Import required modules
+    from toss.trajectory.Integrator import IntegrationScheme
```
Comment (gomezzz) [inline toss/trajectory/compute_trajectory.py:175]: at the top

File: toss/trajectory/compute_trajectory.py [hunk 8]
```diff
@@ -210,53 +198,53 @@ def integrate_system(func: Callable, x: np.ndarray, args):
...
 
     Returns:
-        collision_boolean (bool): A one dimensional array with boolean values corresponding to each
-                                position kept in x. Returns "False" if point is inside mesh, and 
-                                "True" if point is outside mesh (that is, there no collision).
+        collision_boolean (bool): Array with boolean values corresponding to each position in x. 
+                                  Returns "False" if point is inside mesh, and "True" if point is 
+                                  outside mesh (that is, there no collision).
     """
+    from toss.mesh.mesh_utility import is_outside
```
Comment (gomezzz) [inline toss/trajectory/compute_trajectory.py:248]: 🔝

File: toss/trajectory/trajectory_tools.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            states = np.transpose(ode_object.y)
+            timesteps = ode_object.t
+        else:
+            states = np.hstack((states, np.transpose(ode_object.y)))
+            timesteps = np.hstack((timesteps, ode_object.t))
+            
+    return states, timesteps
+
+
+def get_trajectory_fixed_step(args, list_of_ode_objects: list) -> Union[np.ndarray, np.ndarray]:
```
Comment (gomezzz) [inline toss/trajectory/trajectory_tools.py:28]: this function deserves a test if possible to make sure the intervals are as expected

File: toss/fitness/fitness_function_utils.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# Core packages
+import numpy as np
+import typing
+
+def estimate_covered_volume(positions: np.ndarray) -> float:
```
Comment (gomezzz) [inline toss/fitness/fitness_function_utils.py:5]: A test for this function maybe nice

File: toss/fitness/fitness_functions.py
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:1]: you might want to consider simple tests for the fitness functions as sanity checks

**Review by gomezzz — CHANGES_REQUESTED** [review]
Looks a lot better! Some minor formatting comments and you might want to add some simple tests for a few components to be sure they work as intended and keep doing so. almost there :v:

File: toss/fitness/fitness_functions.py [hunk 11]
```diff
@@ -1,98 +1,121 @@
...
 
 
-def TargetAltitudeDistance(args, positions: np.ndarray, timesteps: None) -> float:
+def target_altitude_distance(args, positions: np.ndarray, timesteps: None) -> float:
     """ Computes average distance to target altitude for satellite positions along the trajectory.
     Args:
-        args (dotmap.DotMap): Dotmap with information on target squared altitude of spacecraft.
-        positions (np.ndarray): A set of positions along the trajectory.
-        timesteps (None): Time values for each position. 
+        args (dotmap.DotMap): Dotmap with information on radius of outer bounding sphere.
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:41]: why a dotmap? you only use one float, just pass that? also avoids the very generic `args`

File: toss/fitness/fitness_functions.py [hunk 12]
```diff
@@ -1,98 +1,121 @@
...
 
-def CloseDistancePenalty(args, positions: np.ndarray, timesteps: None) -> float:
-    """ Computes average deviation from the inner-bounding sphere of satellite positions inside the inner bounding-sphere.
+def close_distance_penalty(args, positions: np.ndarray, timesteps: None) -> float:
+    """ Computes penalty depending on the position closest to the body and inside risk-zone.
     Args:
-        args (dotmap.DotMap): Dotmap with information on radius of inner bounding sphere.
-        positions (np.ndarray): A set of positions along the trajectory.
-        timesteps (None): Time values for each position. 
+        args (dotmap.DotMap): Dotmap with information on radius of outer bounding sphere.
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:56]: as above

File: toss/fitness/fitness_functions.py [hunk 13]
```diff
@@ -1,98 +1,121 @@
...
+    penalty = maximum_distance/((maximum_distance + (args.problem.radius_inner_bounding_sphere**2))/2)
+    return penalty
 
 
-def FarDistancePenalty(args, positions: np.ndarray, timesteps: None) -> float:
-    """ Computes average deviation from the outer-bounding sphere of satellite positions outside the outer bounding-sphere.
+def far_distance_penalty(args, positions: np.ndarray, timesteps: None) -> float:
+    """ Computes penalty depending on the position farthest away from measurement-zone.
     Args:
         args (dotmap.DotMap): Dotmap with information on radius of outer bounding sphere.
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:78]: as above

File: toss/fitness/fitness_functions.py [hunk 14]
```diff
@@ -1,98 +1,121 @@
...
     estimated_volume = estimate_covered_volume(positions)
     fitness = -(estimated_volume/args.problem.measurable_volume)
     return fitness
 
 
-def CoveredVolumeFarDistancePenalty(args, positions: np.ndarray, timesteps: np.ndarray) -> float:
+def covered_volume_far_distance_penalty(args, positions: np.ndarray, timesteps: np.ndarray) -> float:
     """ Returns combined fitness value of CoveredVolume and FarDistancePenalty
     Args:
         args (dotmap.DotMap): Dotmap with information on total measurable volume and radius of outer bounding sphere.
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:114]: as above

File: toss/trajectory/compute_trajectory.py [hunk 15]
```diff
@@ -1,7 +1,18 @@
+import sys
+sys.path.append("..")
+sys.path.append("../..")
```
Comment (gomezzz) [inline toss/trajectory/compute_trajectory.py:3]: as in other file, don't mess with user's path 🙏

File: scripts/trajectory_plot.png
Comment (gomezzz) [inline scripts/trajectory_plot.png:1]: not sure if intentionally added?

File: toss/fitness/fitness_functions.py [hunk 16]
```diff
@@ -1,121 +1,160 @@
...
+    distance_squared = _compute_squared_distance(positions, radius_inner_bounding_sphere)
+
     # We only want to penalize positions that are inside inner-sphere (i.e risk-zone).
-    # For positions inside sphere, identify the one farthest away from radius (i.e with greatest risk)
-    maximum_distance = np.abs(np.min(distance_squared[distance_squared<0]))
-    
-    # Determine penalty P=[0,1] depending on distance. 
-    penalty = maximum_distance/((maximum_distance + (args.problem.radius_inner_bounding_sphere**2))/2)
-    return penalty
+    distance_squared = distance_squared[distance_squared<0]
```
Comment (gomezzz) [inline toss/fitness/fitness_functions.py:67]: How can squared numbers be <0? :)

File: toss/trajectory/equations_of_motion.py [hunk 17]
```diff
@@ -73,8 +75,13 @@ def compute_motion(t: float, x: np.ndarray, args) -> np.ndarray:
     Returns:
         (np.ndarray): K vector used for computing state at the following time step.
     """
-    rotated_position = rotate_point(t, x[0:3], args)
-    a = compute_acceleration(rotated_position, args)
+    position = x[0:3]
+
+    if args.problem.activate_rotation == True:
```
Comment (gomezzz) [inline toss/trajectory/equations_of_motion.py:80]:
```suggestion
    if args.problem.activate_rotation:
```

**Review by gomezzz — APPROVED** [review]
just some minor comments, not sure if they need addressing

**Pull request merged**